pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotImplicitTableHintRule.java

```diff
@@ -0,0 +1,316 @@
+            partitionParallelism);
+        return partitionParallelism;
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException(
+            "Invalid partition parallelism: " + partitionParallelismStr + " for table: " + tableScan);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Get the partition size from the hint, if any, otherwise use the partition size from the table partition info.
+   */
+  private static int getPartitionSize(TablePartitionInfo tablePartitionInfo, LogicalTableScan tableScan,
+      Map<String, String> kvOptions) {
+    int partitionSize = tablePartitionInfo.getNumPartitions();
+    String partitionSizeStr = kvOptions.get(PinotHintOptions.TableHintOptions.PARTITION_SIZE);
+    if (partitionSizeStr != null) {
+      try {
+        int explicitPartitionSize = Integer.parseInt(partitionSizeStr);
+        if (explicitPartitionSize != partitionSize) {
+          LOGGER.debug("Override implicit table hint for {} with explicit partition size: {}", tableScan,
+              explicitPartitionSize);
+          partitionSize = explicitPartitionSize;
+        }
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException("Invalid partition size: " + partitionSizeStr + " for table: " + tableScan);
+      }
+    }
+    return partitionSize;
+  }
+
+  /**
+   * Get the table partition info for the given table scan, transforming Calcite's table name to Pinot's table name
+   * and adding the table type if needed.
+   */
+  @Nullable
+  private TablePartitionInfo getTablePartitionInfo(LogicalTableScan tableScan) {
+    String tableName = RelToPlanNodeConverter.getTableNameFromTableScan(tableScan);
+
+    String offlineName = TableNameBuilder.forType(TableType.OFFLINE).tableNameWithType(tableName);
+    TablePartitionInfo offlineTpi = _workerManager.getTablePartitionInfo(offlineName);
+    String realtimeName = TableNameBuilder.forType(TableType.REALTIME).tableNameWithType(tableName);
+    TablePartitionInfo realtimeTpi = _workerManager.getTablePartitionInfo(realtimeName);
+
+    if (offlineTpi == null && realtimeTpi == null) {
+      LOGGER.debug("Table partition info not found for table: {}", tableName);
+      return null;
+    }
+    if (offlineTpi == null) {
+      return realtimeTpi;
+    }
+    if (realtimeTpi == null) {
+      return offlineTpi;
+    }
+    if (!_workerManager.isFullyReplicated(tableName)) {
+      LOGGER.debug("Table {} is not fully replicated", tableName);
+      return null;
+    }
+    // both tpis are equal, so we can return either
+    return offlineTpi;
+  }
+
+  @Value.Immutable
+  public interface Config extends RelRule.Config {
+    @Nullable
+    WorkerManager getWorkerManager();
+
+    @Override
+    default PinotImplicitTableHintRule toRule() {
+      return new PinotImplicitTableHintRule(this);
+    }
+  }
+
+  /**
+   * An interface to find the table partition info for a given table name.
+   */
+  public interface PartitionTableFinder {
```

Comment:
Is this used?

Reply:
No. Now we directly use WorkerManager because we needed to call a couple of methods on that class.